src/index.c

```diff
@@ -2647,30 +2646,33 @@ static int write_disk_entry(git_filebuf *file, git_index_entry *entry, const cha
 	 *
 	 * In 2038 I will be either too dead or too rich to care about this
 	 */
-	ondisk->ctime.seconds = htonl((uint32_t)entry->ctime.seconds);
-	ondisk->mtime.seconds = htonl((uint32_t)entry->mtime.seconds);
-	ondisk->ctime.nanoseconds = htonl(entry->ctime.nanoseconds);
-	ondisk->mtime.nanoseconds = htonl(entry->mtime.nanoseconds);
-	ondisk->dev = htonl(entry->dev);
-	ondisk->ino = htonl(entry->ino);
-	ondisk->mode = htonl(entry->mode);
-	ondisk->uid = htonl(entry->uid);
-	ondisk->gid = htonl(entry->gid);
-	ondisk->file_size = htonl((uint32_t)entry->file_size);
+	ondisk.ctime.seconds = htonl((uint32_t)entry->ctime.seconds);
+	ondisk.mtime.seconds = htonl((uint32_t)entry->mtime.seconds);
+	ondisk.ctime.nanoseconds = htonl(entry->ctime.nanoseconds);
+	ondisk.mtime.nanoseconds = htonl(entry->mtime.nanoseconds);
+	ondisk.dev = htonl(entry->dev);
+	ondisk.ino = htonl(entry->ino);
+	ondisk.mode = htonl(entry->mode);
+	ondisk.uid = htonl(entry->uid);
+	ondisk.gid = htonl(entry->gid);
+	ondisk.file_size = htonl((uint32_t)entry->file_size);
+
+	git_oid_cpy(&ondisk.oid, &entry->id);
 
-	git_oid_cpy(&ondisk->oid, &entry->id);
+	ondisk.flags = htons(entry->flags);
 
-	ondisk->flags = htons(entry->flags);
+	memcpy(mem, &ondisk, sizeof(struct entry_short));
 
 	if (entry->flags & GIT_IDXENTRY_EXTENDED) {
-		struct entry_long *ondisk_ext;
-		ondisk_ext = (struct entry_long *)ondisk;
-		ondisk_ext->flags_extended = htons(entry->flags_extended &
+		struct entry_long ondisk_ext;
+		memcpy(&ondisk_ext, mem, sizeof(struct entry_long));
```

Comment:
Huh. So this doesn't make a lot of sense to me. `mem` was mmapped previously, which is why we assigned that and then just continued writing via the `ondisk_ext` pointer. But what you're doing now is first write `ondisk` to disk via the mmapped file buffer, then copy `ondisk` into `ondisk_ext` and overwrite what you've previously written to the mmapped buffer with `ondisk_ext`. It would make sense to avoid that overhead and only write either `ondisk` or `ondisk_ext` to disk, but not both.

So essentially, I guess what I'm asking for is to move the above `memcpy` into these conditionals here:

1. prepare `struct entry_short`
2. is it extended? copy `entry_short` into `entry_long` and `memcpy` that to `mem`. Otherwise, just `memcpy` `entry_short` into `mem`

Does that make sense to you?